### PR TITLE
Remove jsrsasign

### PR DESCRIPTION
**Pull request by linjie997**
This is still a WIP, but in my local environment it works well.
I have tried login with code flow and refresh tokens, working as expected.
Some unit tests still need to be fixed.

Related to https://github.com/damienbod/angular-auth-oidc-client/issues/506 and https://github.com/damienbod/angular-auth-oidc-client/issues/26.
This would be a pretty big change in the library since we could remove the big dependency from jsrsasign, so some feedback/opinion would be appreciated since I'm not 100% sure about the changes so far.

### CODE REVIEW

**Comment by damienbod** [comment]
@linjie997 Thank for this, I would need to do a lot of testing for this, not sure if this will work. I will have to setup all the supported keys on my test IDP. Won't have time for the near future, got a lot happening in the next week. Will get around to this then.

We would need to test this against all the RSA and ECDSA certificates.

Greetings Damien

**Comment by linjie997** [comment]
Sure, let's keep in touch!

**Comment by d-koppenhagen** [comment]
Hey! Is there any roadmap when this feature/refactoring will be released? 
Will it be a breaking change, as the API isn't implemented in older (legacy) browsers? 
Probably a hint for using a polyfill would be good wouldn't it?

**Comment by linjie997** [comment]
Hello @d-koppenhagen, from my perspective it should be working and just need to be tested as damien said.
About the legacy browsers (I guess you mean IE), this implementation uses [Crypto.subtle](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/subtle), so if IE support is needed maybe [this polyfill](https://github.com/PeculiarVentures/webcrypto-liner) might work.

**Comment by damienbod** [comment]
I hope to get around to testing this soon, I have a few different key types to test this.

**Comment by FabianGosebrink** [comment]
Hey, this is great work. Can you fix the unit tests? Thanks

**Comment by linjie997** [comment]
I'll try to fix them as soon as I find some time, I'm currently on vacation :)

**Comment by FabianGosebrink** [comment]
That is absolutely more important than coding! Enjoy the vacation 👋 ☀⛱🌊

**Comment by linjie997** [comment]
@FabianGosebrink ehy, unit tests should have been all fixed.
I have been a bit busy with vacation and new job so it took a while :)

**Comment by FabianGosebrink** [comment]
Thanks for this. Can you fix the linting please?

**Comment by linjie997** [comment]
Linting should be fine now!

**Comment by damienbod** [comment]
Hi @linjie997 Let me know if I can help, I will invest time here, see if I can solve this as well,

Greetings Damien

**Comment by linjie997** [comment]
Hi @damienbod, if you have time to take a look into it and give some help I would be really grateful! Unfortunately I have been very busy in RL

**Comment by linjie997** [comment]
@FabianGosebrink Hello Fabian, thank you for the code review.
I think you saw an older version of the code, with the latest commit I moved to observables instead of promises

**Comment by FabianGosebrink** [comment]
Oh I am sorry! I will review again tomorrow and use the latest code then. Thanks for your effort! Would love to get this merged. @damienbod is currently testing iirc. Thanks 👍

**Comment by linjie997** [comment]
Greetings!
I finally had some time and I pushed some fix for various keys and conflicts from main branch.
My tests passed and it would be great to have some feedback from you @damienbod.
Sorry for taking so long and hope to hear you soon.

**Comment by damienbod** [comment]
@linjie997 awesome, testing now, thanks

**Comment by linjie997** [comment]
Some unit test are failing at the moment, if the code itself looks good and keys test are passing I will work on the unit testing too, there are quite a bit to refactor :)

File: projects/angular-auth-oidc-client/package.json
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/package.json:13]: remove

File: projects/angular-auth-oidc-client/package.json
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/package.json:14]: remove

File: projects/angular-auth-oidc-client/ng-package.json
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/ng-package.json:6]: remove the umdModuleIds please

File: projects/angular-auth-oidc-client/src/lib/flows/callback-handling/state-validation-callback-handler.service.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/flows/callback-handling/state-validation-callback-handler.service.ts:24]: Change returntype to `: Observable<CallbackContext | Observable<never>>`

File: projects/angular-auth-oidc-client/src/lib/flows/callback-handling/state-validation-callback-handler.service.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/flows/callback-handling/state-validation-callback-handler.service.ts:39]: remove this throwError as it is a merge mistake :)

**Comment by FabianGosebrink** [comment]
I did a few comments because we could not build and run the lib. Please have a look, thanks!

**Comment by linjie997** [comment]
Updated, should be fine now

**Comment by damienbod** [comment]
Thanks @linjie997 @FabianGosebrink tested the functionality on Firefox, Chrome and Edge

All EC and RSA work on Edge and Chrome

All RSA work on Firefox

 ECDsaSigningAlgorithm.ES256 and ES384 don't work on Firefox => this worked before, something small wrong here and only just for Firefox.

Maybe you see something, looking as well.

Greetings Damien

**Comment by damienbod** [comment]
@linjie997 

I think this is the problem:

https://bugzilla.mozilla.org/show_bug.cgi?id=1718899

I think if the key EC, we need to remove the alg value so that firefox works, no idea how this effects the Chrome and the Edge implementations, testing now

would need some type of fix here:

```
  return from(this.cyptoObj.subtle.importKey('jwk', key, algorithm, false, ['verify'])).pipe(
      mergeMap((cryptoKey: CryptoKey) => {
        const signature: Uint8Array = base64url.parse(rawSignature, {loose: true});

        const algorithm: RsaHashedImportParams | EcdsaParams = this.getVerifyAlg(alg);
        return from(this.cyptoObj.subtle.verify(algorithm, cryptoKey, signature, new TextEncoder().encode(signingInput)));
      }),
      tap((isValid: boolean) => {
        if (!isValid) {
          this.loggerService.logWarning(configId, 'incorrect Signature, validation failed for id_token');
        }
      })
    );
  }
```

**Comment by damienbod** [comment]
This fixes the Firefox bug

```
const agent = window.navigator.userAgent.toLowerCase();

if (agent.indexOf('firefox') > -1 && key.kty === 'EC') {
  key.alg = '';
}

```

**Comment by linjie997** [comment]
Great finding, I will update the code later with this addition

**Comment by linjie997** [comment]
Updated with Firefox hack.
Seems that everything is working good, I will dedicate some time in the coming days to unit test fixes :)

**Comment by damienbod** [comment]
This works now, really awesome, massive thanks!

All we need to now is a code clean up and code quality. We probably could rename the **JsrsAsignReducedService** service. And the file names to **jsrsasign-reduced.service.**

could you remove this from the package.json : "jsrsasign-reduced": "^8.0.15",

@FabianGosebrink we will release this as version 13.1.0 (before any other PRs , good?)

Greetings Damien

**Comment by FabianGosebrink** [comment]
Yes, if the code builds and the tests are running I will do a review.

**Comment by linjie997** [comment]
Greetings. I have fixed the majority of unit test, there is still one failing but I'm not sure how to fix it.
It's in the TokenValidationService, do you perhaps have time to take a look at it?
A second point would be how to rename the JsrsAsignReducedService, since it is not using JsrsAsign anymore.

Thanks!

**Comment by FabianGosebrink** [comment]
I just improved the tests a bit and saw that you are using `async` for the tests. In Angular, there is no `async` to wait for a test to complete all its asynchronous operations anymore. It is all `waitForAsync`. The tests now run, but the one fails:

```
TokenValidationService validateIdTokenAtHash returns true if sha is sha256 and generated hash equals atHash param FAILED
        Error: Expected false to equal true.
``` 

I don't know about the encryption, so maybe you can have a look, or @damienbod can help. 

@linjie997 Can you please replace all `async` in the tests with `waitForAsync`? Thanks.

**Comment by linjie997** [comment]
Hello,
I removed all unused `async`.
Abou the failing unit test, I think it was a merge mistake and I reverted to the old `atHash` value. Now the test is passing :)

**Comment by FabianGosebrink** [comment]
Please do Singlequotes everywhere, no double quotes.

File: .prettierrc
Comment (FabianGosebrink) [inline .prettierrc:5]: We don't need this, do we?

File: projects/angular-auth-oidc-client/src/lib/flows/callback-handling/state-validation-callback-handler.service.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/flows/callback-handling/state-validation-callback-handler.service.ts:38]: Please use rxjs operator: return throwError(() => new Error(errorMessage));

File: projects/angular-auth-oidc-client/src/lib/iframe/refresh-session-iframe.service.spec.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/iframe/refresh-session-iframe.service.spec.ts:37]: Can this be deleted? Or do we need it and it needs to be taken a look at?

File: projects/angular-auth-oidc-client/src/lib/iframe/refresh-session-iframe.service.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/iframe/refresh-session-iframe.service.ts:25]: This returns an observable, doesnt it?

File: projects/angular-auth-oidc-client/src/lib/login/par/par.service.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/login/par/par.service.ts:35]: no `from` needed, `createBodyForParCodeFlowRequest` returns an observable.

File: projects/angular-auth-oidc-client/src/lib/login/popup/popup-login.service.spec.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/login/popup/popup-login.service.spec.ts:1]: Singlequotes

File: projects/angular-auth-oidc-client/src/lib/utils/tokenHelper/token-helper.service.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/utils/tokenHelper/token-helper.service.ts:29]: I like this, good job!

File: projects/angular-auth-oidc-client/src/lib/utils/url/url.service.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/utils/url/url.service.ts:404]: lets return an empty string here.

File: projects/angular-auth-oidc-client/src/lib/utils/url/url.service.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/utils/url/url.service.ts:463]: lets return an empty string here

File: projects/angular-auth-oidc-client/src/lib/validation/token-validation.service.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/validation/token-validation.service.ts:497]: remove console.warn

**Review by FabianGosebrink — CHANGES_REQUESTED** [review]

File: projects/angular-auth-oidc-client/src/lib/flows/callback-handling/state-validation-callback-handler.service.ts
Comment (linjie997) [reply projects/angular-auth-oidc-client/src/lib/flows/callback-handling/state-validation-callback-handler.service.ts:38]: Using `return throwError(() => new Error(errorMessage));` would make the method return `Observable<CallbackContext | Observable<never>>` and cause errors in `FlowsService`, do you think it's the best approach? We would need some checks and rewrite in FlowsService.

Comment (FabianGosebrink) [reply projects/angular-auth-oidc-client/src/lib/flows/callback-handling/state-validation-callback-handler.service.ts:38]: Nice catch. I skipped the `map` operator for some reason. Okay then, `throw new Error()` is okay then, just add a new line before.

Comment (linjie997) [reply projects/angular-auth-oidc-client/src/lib/flows/callback-handling/state-validation-callback-handler.service.ts:39]: Done :)

File: projects/angular-auth-oidc-client/src/lib/utils/url/url.service-mock.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/utils/url/url.service-mock.ts:49]: Observable here

File: projects/angular-auth-oidc-client/src/lib/validation/token-validation.service-mock.ts
Comment (FabianGosebrink) [reply projects/angular-auth-oidc-client/src/lib/validation/token-validation.service-mock.ts:42]: Observable

File: projects/angular-auth-oidc-client/src/lib/validation/token-validation.service-mock.ts
Comment (FabianGosebrink) [reply projects/angular-auth-oidc-client/src/lib/validation/token-validation.service-mock.ts:34]: Observable

File: projects/angular-auth-oidc-client/src/lib/login/standard/standard-login.service.spec.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/login/standard/standard-login.service.spec.ts:1]: Singlequotes all over

File: projects/angular-auth-oidc-client/src/lib/validation/state-validation.service.spec.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/validation/state-validation.service.spec.ts:105]: waitForASync

**Review by FabianGosebrink — CHANGES_REQUESTED** [review]

**Comment by FabianGosebrink** [comment]
Please do `waitForAsync` all over and Single quotes all over. Thanks!

**Comment by linjie997** [comment]
Should be good now, I checked the changes one more time

**Comment by linjie997** [comment]
Should be good now, I checked the changes one more time

File: projects/angular-auth-oidc-client/src/lib/utils/url/url.service-mock.ts
Comment (FabianGosebrink) [inline projects/angular-auth-oidc-client/src/lib/utils/url/url.service-mock.ts:53]: No async, no Promise, just Observable here

**Review by FabianGosebrink — CHANGES_REQUESTED** [review]

**Comment by FabianGosebrink** [comment]
Cool! We are getting there. Just one small comment. @damienbod can you do a last test run before merge?

**Comment by damienbod** [comment]
I do a test then, thanks @linjie997 @FabianGosebrink

**Comment by linjie997** [comment]
Great, it took a while but I'm glad it worked out good at the end.
I'll wait for the last tests feedback!

**Comment by damienbod** [comment]
@linjie997 Thanks for this, this was a huge PR and took so long. Thanks so much for you patience and effort

preparing the release now 13.1.0

Greetings Damien

**Comment by linjie997** [comment]
Great news, happy that I contribuited!!

**Comment by FabianGosebrink** [comment]
Congratulations! Great one!

**Comment by jeroenheijmans** [comment]
I'd been following this PR with great interest, and at the risk of posting a "+1" style comment I wanted to explicitly thank y'all for this hard work in creating, reviewing, and completing this PR! This work will serve the ecosystem, the devs, but even more so many consumers (and their data plans 😅) downstream. Great work!

**Comment by damienbod** [comment]
Thanks @jeroenheijmans

**Comment by FabianGosebrink** [comment]
Thanks @jeroenheijmans !

File: projects/angular-auth-oidc-client/src/lib/validation/jsrsasign-reduced.service.ts
Comment (GabrielGil) [inline projects/angular-auth-oidc-client/src/lib/validation/jsrsasign-reduced.service.ts:7]: This (and the other global `window` calls) break support of Angular Universal (SSR).

The correct way to use window, is by injecting the document `@Inject(DOCUMENT) private readonly doc: Document,` and then using it `this.doc.defaultView`.

Also, since this version is for Angular 13, and the support  of Internet Explorer is deprecated by the Angular team, i don't think it makes sense to consider IE11.

**Review by GabrielGil — COMMENTED** [review]
I suggest that `window` is never accessed from the global namespace, as it may not be available, for instance in the server, and the injection of the document should be used instead.

Refer to #913 for more.

File: projects/angular-auth-oidc-client/src/lib/validation/jsrsasign-reduced.service.ts
Comment (FabianGosebrink) [reply projects/angular-auth-oidc-client/src/lib/validation/jsrsasign-reduced.service.ts:7]: That is absolutely correct. We already use the injection of the document in other places. We will fix that in a separate PR. Thanks for the hint!

Comment (GabrielGil) [reply projects/angular-auth-oidc-client/src/lib/validation/jsrsasign-reduced.service.ts:7]: No problem. We collaborate to help each other :) Thanks to you guys!

Comment (FabianGosebrink) [reply projects/angular-auth-oidc-client/src/lib/validation/jsrsasign-reduced.service.ts:7]: https://github.com/damienbod/angular-auth-oidc-client/issues/1316

**Comment by Budno** [comment]
@FabianGosebrink Could you mark this as a breaking change? Because due to #1320 this ends the support for http.

**Comment by FabianGosebrink** [comment]
Why is it ending the HTTP support? We still use HTTP to communicate.

**Comment by DamienBowden** [comment]
@Budno  @FabianGosebrink yes https is required to use the browser APIs

**Comment by Lexa-tech** [comment]
Hi @linjie997 , did you try the polyfill webcrypto-liner you mentionned here when using HTTP ?
https://github.com/damienbod/angular-auth-oidc-client/pull/1220#issuecomment-908551547

**Comment by IlyaSurmay** [comment]
> Hi @linjie997 , did you try the polyfill webcrypto-liner you mentionned here when using HTTP ? [#1220 (comment)](https://github.com/damienbod/angular-auth-oidc-client/pull/1220#issuecomment-908551547)

I tried it and it didn't seem to work
image
adding asmcrypto library didn't help much, but the error message changed to `ERROR TypeError: asmCrypto.SHA256 is undefined`

**Comment by desmondblue** [comment]
I went through the conversation and saw @linjie997  had pushed some fixes to version 13.x version.  Hence my question - Does that imply the breaking change that @Budno mentions "browser APIs will only use https" is starting from 13.x while the 12.x plugin may work fine without the crypto.digest issue?

**Pull request merged**
